addons/product/product.py

```diff
@@ -1069,26 +1069,26 @@ def name_search(self, cr, user, name='', args=None, operator='ilike', context=No
             positive_operators = ['=', 'ilike', '=ilike', 'like', '=like']
             ids = []
             if operator in positive_operators:
-                ids = self.search(cr, user, [('default_code','=',name)]+ args, limit=limit, context=context)
+                ids = self.search(cr, user, ['|', ('default_code', '=', name), ('attribute_value_ids.name', operator, name)] + args, limit=limit, context=context)
                 if not ids:
-                    ids = self.search(cr, user, [('barcode','=',name)]+ args, limit=limit, context=context)
+                    ids = self.search(cr, user, [('barcode', '=', name)] + args, limit=limit, context=context)
```

Comment:
Please don't modify the search by `barcode`, this must be another separate, independent exact match done before everything else.